Add Node.js executor for DMOJ

pxpeterxu opened the pull request:
Thanks again for maintaining DMOJ! It's a great project with a really solid security foundation.

### Motivation

We wanted to use DMOJ for internal company events, and since our company's primary language is Javascript (using Node.js), we wanted to implement a version that used Node directly rather than using https://github.com/DMOJ/v8dmoj

This resolves issue https://github.com/DMOJ/judge-server/issues/1136

### Changes

- Add a judge executor that will execute code written in Node.js
  - This was based on the Coffeescript executor, except with some newer syscalls added (e.g., io_uring)

#### Related PRs on other repositories

- https://github.com/DMOJ/runtimes-docker/pull/71: Update Dockerfile to install Node.js to tier3 runtime image

### Testing

Deployed a judge server with this with the following Dockerfile:

```sh
FROM dmoj/runtimes-tier3

# Install Node from nodesource based on https://github.com/nodesource/distributions/tree/69a4558#installation-instructions
RUN apt-get install -y ca-certificates curl gnupg && mkdir -p /etc/apt/keyrings && { curl -fsSL https://deb.nodesource.com/gpgkey/nodesource-repo.gpg.key | gpg --dearmor -o /etc/apt/keyrings/nodesource.gpg; } && { echo 'deb [signed-by=/etc/apt/keyrings/nodesource.gpg] https://deb.nodesource.com/node_20.x nodistro main' > /etc/apt/sources.list.d/nodesource.list; } && apt-get update && apt-get install -y nodejs

# The remainder of this file is copied from
# https://github.com/DMOJ/judge-server/blob/master/.docker/tier3/Dockerfile
RUN mkdir /judge /problems && cd /judge && \
	curl -L https://github.com/wanderlog/judge-server/archive/master.tar.gz | tar -xz --strip-components=1 && \
    python3 -m venv --prompt=DMOJ /env && \
	/env/bin/pip3 install cython && \
	/env/bin/pip3 install -e . && \
	/env/bin/python3 setup.py develop && \
	HOME=~judge . ~judge/.profile && \
	runuser -u judge -w PATH -- /env/bin/dmoj-autoconf -V > /judge-runtime-paths.yml && \
	echo '  crt_x86_in_lib32: true' >> /judge-runtime-paths.yml
```

Verified that test runs using this Dockerfile worked properly:

![Screen capture](https://github.com/DMOJ/judge-server/assets/2924388/053252d5-b0ee-48d4-8b1d-e46957e43b31)

dmoj-build commented:
Can one of the admins verify this patch?

quantum5 commented:
Your code looks fairly good, but I have to ask: what's the point of this? If you look at your submission for the A Plus B problem, you'll note that you are doing a lot of I/O boilerplate, mostly because node.js is designed for stream-based I/O and not so much for line-based I/O. You also won't be able to use `npm` modules on the judge.

The V8JS runtime is the exact same JS engine except it comes with a bunch of line-based I/O functions that requires a lot less boilerplate to use.

pxpeterxu commented:
Good point -- I think the main goal of this was for our engineers to run this without having to install the https://github.com/DMOJ/v8dmoj client locally, since Node.js is already preinstalled on their machines and is pretty widely used! It does require somewhat more boilerplate in code, though we often provide this with the problem.

I'd understand if this is too different from the goals of the main DMOJ site though; let me know what might be helpful.

quantum5 commented on dmoj/executors/NODEJS.py:
```diff
@@ -0,0 +1,44 @@
+        'statx',
+        # End syscalls copied from COFFEE.py
+        
+        # Without this, we get
+        # > Protection fault on: 425 (sys_io_uring_setup)
```
This comment is not very helpful.
```suggestion
```

quantum5 commented on dmoj/executors/NODEJS.py:
```diff
@@ -0,0 +1,44 @@
+    nproc = -1
+    command_paths = ['node', 'nodejs']
+    syscalls = [
+        # Start syscalls copied from COFFEE.py
+        'capget',
+        'eventfd2',
+        'newselect',
+        'sched_yield',
+        'select',
+        'setrlimit',
+        'statx',
+        # End syscalls copied from COFFEE.py
```
Can we try trimming down this list? Also the comment is not necessary.

pxpeterxu replied:
Thanks for the suggestion! I ran `/judge/.docker/entry test | grep NODEJS` multiple times while commenting out these, and it seems like we only need `capget` and `eventfd2`. I left these two (and `shutdown`) in.

quantum5 commented on dmoj/executors/NODEJS.py:
```diff
@@ -0,0 +1,44 @@
+        # Without this, we get
+        # > Protection fault on: 425 (sys_io_uring_setup)
+        'io_uring_setup',
+        # Without this, when running an actual test case (though not the
+        # self-test), we get
+        # > RTE (shutdown syscall disallowed)
```
This comment is not very helpful.
```suggestion
```

quantum5 commented on dmoj/executors/NODEJS.py:
```diff
@@ -0,0 +1,44 @@
+        
+        # Without this, we get
+        # > Protection fault on: 425 (sys_io_uring_setup)
+        'io_uring_setup',
```
Can we try making this return `ENOSYS`? I don't think `io_uring` at this time due to potential security implications.

pxpeterxu replied:
Good call: I was able to remove this by setting an environment variable to disable this:

```
    def get_env(self):
        env = super().get_env()
        # Disable io_uring due to potential security implications
        env['UV_USE_IO_URING'] = '0'
        return env
```

quantum5 reviewed (commented):
After some discussion, we've decided that Node.js is an okay addition after all.

pxpeterxu commented:
@quantum5 Thanks again for the review and happy new year! I think I've addressed all comments; can you please take a look?

Also, if you can check the associated PRs, that'd be great too:

- https://github.com/DMOJ/runtimes-docker/pull/71
- https://github.com/DMOJ/online-judge/pull/2297
- https://github.com/DMOJ/docs/pull/155

Xyene commented on README.md:
```diff
@@ -66,6 +66,7 @@ The judge can also grade in the languages listed below:
 * Groovy
 * Haskell
 * INTERCAL
+* JavaScript (Node.js or V8)
```
For consistency with C++ above.

```suggestion
* JavaScript (Node.js and V8)
```

pxpeterxu replied:
Changed it to "JavaScript (Node.js and V8)"

Xyene commented:
It looks like lint failed, but otherwise this looks good to me.

Once you've resolved that, could you please rebase and squash your commits? After that we should be good to merge.

murtadha commented:
I'm eagerly waiting for this PR to be merged as our community has also been asking for the ability to run JS code via Node.js for exactly the same reason @pxpeterxu mentioned.

Please let me know if there is anything I can do to help get this merged ASAP 🙏

int-y1 commented on dmoj/executors/NODEJS.py:
```diff
@@ -0,0 +1,36 @@
+from dmoj.executors.script_executor import ScriptExecutor
+from dmoj.cptbox.filesystem_policies import ExactFile
```
the linter says that import statements are in the wrong order

pxpeterxu replied:
Thanks; I did `pip3 install flake8 flake8-black` and made sure this works now

int-y1 commented on dmoj/executors/NODEJS.py:
```diff
@@ -0,0 +1,36 @@
+    command = 'node'
+    nproc = -1
+    command_paths = ['node', 'nodejs']
+    syscalls = [
```
the linter doesn't like the formatting here

try changing `syscalls` to fit on 1 line

pxpeterxu replied:
Changed, and it seems like it passes `black` now

pxpeterxu commented:
@int-y1 I fixed the lint errors! Hopefully this is ready to merge now, but let me know if there are any other changes (sorry, it's been a busy few weeks!)

Xyene commented:
Thanks, could you please rebase and squash your commits into one?

pxpeterxu commented:
> Thanks, could you please rebase and squash your commits into one?

Just done! Also, I didn't realize this project started at Don Mills C.I. -- I went there way back too! (Graduated 2010)

Xyene commented:
Wow, small world!

Xyene commented:
Looks like this failed CI because of

```
WARNING:dmoj.cptbox:Protection fault on: 330 (sys_pkey_alloc)
WARNING:dmoj.cptbox:Arg0: 0x0000000000000000
WARNING:dmoj.cptbox:Arg1: 0x0000000000000001
WARNING:dmoj.cptbox:Arg2: 0x0000000000000001
WARNING:dmoj.cptbox:Arg3: 0x0000150daed57188
WARNING:dmoj.cptbox:Arg4: 0x0000000000000000
WARNING:dmoj.cptbox:Arg5: 0x0000000000000000
```

We should be able to just allow it: https://man7.org/linux/man-pages/man7/pkeys.7.html

pxpeterxu commented:
> Looks like this failed CI because of
> 
> ```
> WARNING:dmoj.cptbox:Protection fault on: 330 (sys_pkey_alloc)
> WARNING:dmoj.cptbox:Arg0: 0x0000000000000000
> WARNING:dmoj.cptbox:Arg1: 0x0000000000000001
> WARNING:dmoj.cptbox:Arg2: 0x0000000000000001
> WARNING:dmoj.cptbox:Arg3: 0x0000150daed57188
> WARNING:dmoj.cptbox:Arg4: 0x0000000000000000
> WARNING:dmoj.cptbox:Arg5: 0x0000000000000000
> ```
> 
> We should be able to just allow it, seems like it's part of some CFI protection in V8:
> 
>     * https://man7.org/linux/man-pages/man7/pkeys.7.html
> 
>     * https://docs.google.com/document/d/1O2jwK4dxI3nRcOJuPYkonhTkNQfbmwdvxQMyXgeaRHo/edit#heading=h.bvaojj9fu6hc

I added a whitelist here, and it seems to be working: https://github.com/wanderlog/judge-server/actions/runs/7618060649

Force-pushed an update to this branch too

Xyene reviewed (approved):
LGTM, thanks!

yanngv29 commented:
not sure if this is the same problem than 5 day ago  but i have got error  today  with dmoj-autoconf: 

Auto-configuring NODEJS: Protection fault on: 331 (sys_pkey_free)
Arg0: 0x0000000000000001
Arg1: 0x0000000000000001
Arg2: 0x0000000000000008
Arg3: 0xe74cfd037f6ee7e7
Arg4: 0x0000000000000007
Arg5: 0x000055555563db90

Xyene commented:
Thanks for the report, I've put together https://github.com/DMOJ/judge-server/pull/1164.
